application/actions/auth.go

```diff
@@ -1,131 +1,199 @@
+type AuthError struct {
+	Code    string `json:"Code"`
+	Message string `json:"Message"`
+}
 
-	clientID := c.Param("client_id")
-	if clientID == "" {
-		return fmt.Errorf("client_id is required to login")
-	}
-	c.Session().Set("ClientID", clientID)
+type AuthOrgOption struct {
+	ID      string `json:"ID"`
+	Name    string `json:"Name"`
+	LogoURL string `json:"LogoURL"`
+}
 
-	returnTo := c.Param("ReturnTo")
-	if returnTo == "" {
-		returnTo = "/"
-	}
-	c.Session().Set("ReturnTo", returnTo)
+type AuthUser struct {
+	ID                   string          `json:"ID"`
+	Name                 string          `json:"Name"`
+	Nickname             string          `json:"Nickname"`
+	Email                string          `json:"Email"`
+	Organizations        []AuthOrgOption `json:"Organizations"`
+	AccessToken          string          `json:"AccessToken"`
+	AccessTokenExpiresAt int64           `json:"AccessTokenExpiresAt"`
+}
 
-	err := c.Session().Save()
-	if err != nil {
-		return err
+type AuthResponse struct {
+	Error          *AuthError       `json:"Error,omitempty"`
+	AuthOrgOptions *[]AuthOrgOption `json:"AuthOrgOptions,omitempty"`
+	RedirectURL    string           `json:"RedirectURL,omitempty"`
+	User           *AuthUser        `json:"User,omitempty"`
+}
+
+func AuthLogin(c buffalo.Context) error {
+	var authEmail string
+	authEmail = c.Param("authEmail")
+
+	var org models.Organization
+	if authEmail == "" {
+		// TODO: find a way around this hack without using session storage
+		var err error
+		if org, err = models.OrganizationFindByDomain("sil.org"); err != nil {
+			return authError(c, http.StatusBadRequest, "MissingAuthEmail", "authEmail is required to login")
+		}
+	} else {
+		// find org for auth config and processing
+		var orgID int
+		oid := c.Param("org_id")
+		if oid == "" {
+			orgID = 0
+		} else {
+			orgID, _ = strconv.Atoi(oid)
+		}
+		userOrgs, err := models.UserOrganizationFindByAuthEmail(authEmail, orgID)
+		if len(userOrgs) == 1 {
+			org = userOrgs[0].Organization
+		}
+
+		// no user_organization records yet, see if we have an organization for user's email domain
+		if len(userOrgs) == 0 {
+			org, err = models.OrganizationFindByDomain(domain.EmailDomain(authEmail))
+			if err != nil {
+				return authError(c, http.StatusInternalServerError, "UnableToFindOrgByEmail", "unable to find organization by email domain")
+			}
+			if org.AuthType == "" {
+				return authError(c, http.StatusNotFound, "OrgNotFound", "unable to find organization by email domain")
+			}
+		}
 	}
 
-	sp, err := getSAML2Provider()
+	// get auth provider for org to process login
+	sp, err := org.GetAuthProvider()
 	if err != nil {
-		return err
+		return authError(c, http.StatusInternalServerError, "UnableToLoadAuthProvider", "unable to load auth provider for organization")
 	}
 
-	authURL, err := sp.BuildAuthURL("")
-
-	return c.Redirect(302, authURL)
-}
-
-func AuthCallback(c buffalo.Context) error {
+	authResp := sp.Login(c)
```

Comment:
I'm not sure what the reason for combining the two original Auth functions was.

In any case, could we move the rest of this function's code to another function that is named to clarify that it only gets called after authn has occurred?

Reply:
I like this suggestion. There is certainly room for cleanup here. I'd like to finish out this feature branch and start fresh with a new branch for refactoring.